### PR TITLE
GO-3170 check objects and files are synced with responsible peers and send sync status event

### PR DESCRIPTION
**Pull request by AnastasiaShemyakinskaya**
https://linear.app/anytype/issue/GO-3170/check-objects-and-files-are-synced-with-responsible-peers-in

Added SpaceSyncStatus service, which reads events from queue and send them to client

### CODE REVIEW

File: core/syncstatus/spacesyncstatus/spacestatus.go [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+type NetworkConfig interface {
+	GetNetworkMode() pb.RpcAccountNetworkMode
+}
+
+type spaceSyncStatus struct {
+	eventSender   event.Sender
+	networkConfig NetworkConfig
+	batcher       *mb.MB[*helpers.SpaceSync]
+
+	filesState   State
```
Comment (deff7) [inline core/syncstatus/spacesyncstatus/spacestatus.go:42]: maybe use maps?

spaceId => State
`filesState map[string]State`

File: core/syncstatus/helpers/helpers.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
+package helpers
```
Comment (deff7) [inline core/syncstatus/helpers/helpers.go:1]: maybe move this to `domain/syncstatus` package?

File: core/syncstatus/objectsyncstatus/syncstatus.go [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+	s.treeHeads[treeId] = treeHeadsEntry{
+		heads:        headsCopy,
+		stateCounter: s.stateCounter,
+		syncStatus:   StatusNotSynced,
+	}
+	s.stateCounter++
+}
+
+func (s *syncStatusService) update(ctx context.Context) (err error) {
+	s.treeStatusBuf = s.treeStatusBuf[:0]
```
Comment (deff7) [inline core/syncstatus/objectsyncstatus/syncstatus.go:142]: seems it should be under lock

Comment (deff7) [reply core/syncstatus/objectsyncstatus/syncstatus.go:142]: got it

File: core/syncstatus/objectsyncstatus/syncstatus.go [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+	HeadsReceive(senderId, treeId string, heads []string)
+}
+
+type StatusWatcher interface {
+	Watch(treeId string) (err error)
+	Unwatch(treeId string)
+	SetUpdateReceiver(updater UpdateReceiver)
+}
+
+type StatusService interface {
```
Comment (deff7) [inline core/syncstatus/objectsyncstatus/syncstatus.go:53]: tests

Comment (AnastasiaShemyakinskaya) [reply core/syncstatus/objectsyncstatus/syncstatus.go:53]: Done

File: core/syncstatus/objectsyncstatus/syncstatus.go [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+	curTreeHeads.heads = slice.DiscardFromSlice(curTreeHeads.heads, func(h string) bool {
+		return h == ""
+	})
+	if len(curTreeHeads.heads) == 0 {
+		curTreeHeads.syncStatus = StatusSynced
+	}
+	s.treeHeads[treeId] = curTreeHeads
+}
+
+func (s *syncStatusService) Watch(treeId string) (err error) {
```
Comment (mcrakhman) [inline core/syncstatus/objectsyncstatus/syncstatus.go:195]: It is important to clear statuses of objects when we call SyncAll with responsible peers (before the method was called something like `RemoveAllExcept...`)

**Review by mcrakhman — APPROVED** [review]

**Comment by requilence** [comment]
need to rework Reload later so it will send events only for specific Session

**Review by requilence — APPROVED** [review]

**Review by mcrakhman — APPROVED** [review]

**Pull request merged**